Enable model process for macOS

mwyrzykowski opened the pull request:
#### 9056258a1f2ac1316b2eea5c855854ff38b74950
<pre>
Enable model process for macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=280096">https://bugs.webkit.org/show_bug.cgi?id=280096</a>
<a href="https://rdar.apple.com/136238786">rdar://136238786</a>

Reviewed by NOBODY (OOPS!).

This doesn&apos;t perform any actual rendering, but it does start building
model process on macOS to ensure we don&apos;t regress code paths for macOS
on that platform related to Model process.

It partially implements iOS as well. iOS requires a few extra changes,
specifically support for a Model Process extension and maybe something else.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
* Source/WebKit/ModelProcess/ios/ModelProcessIOS.mm:
(WebKit::ModelProcess::initializeProcess):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::processStorageClass):
(WebKit::sandboxDirectory):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::launchWithExtensionKitFallback):
(WebKit::launchWithExtensionKit):
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _createVisibilityPropagationView]):
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9056258a1f2ac1316b2eea5c855854ff38b74950

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71421 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50834 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75531 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22626 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73536 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58634 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22445 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56415 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14883 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74487 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46145 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61529 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36855 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42808 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18994 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20967 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64546 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64707 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19360 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77251 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70670 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15655 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18533 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64128 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15698 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61569 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64135 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12274 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5906 "Found 1 new test failure: imported/w3c/web-platform-tests/preload/preload-type-match.html (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92456 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46634 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1413 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20386 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47705 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48988 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47447 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eddydas commented on Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
```diff
@@ -225,6 +225,8 @@ constexpr ASCIILiteral processStorageClass(WebCore::AuxiliaryProcessType type)
 #if ENABLE(GPU_PROCESS)
     case WebCore::AuxiliaryProcessType::GPU:
         return "WebKitGPUSandbox"_s;
+    case WebCore::AuxiliaryProcessType::Model:
```
You may want to guard this with MODEL_PROCESS instead of GPU_PROCESS

eddydas commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -88,6 +88,8 @@
 #if ENABLE(GPU_PROCESS)
     case ProcessLauncher::ProcessType::GPU:
         return { "com.apple.WebKit.GPU"_s, @"com.apple.WebKit.GPU" };
+    case ProcessLauncher::ProcessType::Model:
```
You may want to guard this with MODEL_PROCESS instead of GPU_PROCESS

eddydas commented on Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
```diff
@@ -25,7 +25,7 @@
 #import <Foundation/Foundation.h>
 
-#if defined(TARGET_OS_VISION) && TARGET_OS_VISION
+#if ((defined(TARGET_OS_VISION) && TARGET_OS_VISION) || (defined(TARGET_OS_IOS) && TARGET_OS_IOS) || (defined(TARGET_OS_MAC) && TARGET_OS_MAC))
```
I wonder if MODEL_ELEMENT (not MODEL_PROCESS) is a good flag to use here?

eddydas replied:
or actually maybe MODEL_PROCESS is also a good proxy for this

achan00 commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -221,6 +221,7 @@ @implementation WKContentView {
 #if USE(EXTENSIONKIT)
     RetainPtr<UIView> _visibilityPropagationViewForWebProcess;
     RetainPtr<UIView> _visibilityPropagationViewForGPUProcess;
+    RetainPtr<UIView> _visibilityPropagationViewForModelProcess;
```
Should this be guarded by `MODEL_PROCESS`?

achan00 commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -989,7 +990,7 @@ - (WKVisibilityPropagationView *)_createVisibilityPropagationView
 #endif
 #if ENABLE(MODEL_PROCESS)
-    [self _setupVisibilityPropagationViewForModelProcess];
+    [self _setupVisibilityPropagationForModelProcess];
```
Where is this method implemented?

mwyrzykowski replied:
WKContentView.mm:421

mwyrzykowski committed:
Enable model process for macOS
https://bugs.webkit.org/show_bug.cgi?id=280096
rdar://136238786

Reviewed by NOBODY (OOPS!).

This doesn't perform any actual rendering, but it does start building
model process on macOS to ensure we don't regress code paths for macOS
on that platform related to Model process.

It partially implements iOS as well. iOS requires a few extra changes,
specifically support for a Model Process extension and maybe something else.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
* Source/WebKit/ModelProcess/ios/ModelProcessIOS.mm:
(WebKit::ModelProcess::initializeProcess):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::processStorageClass):
(WebKit::sandboxDirectory):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::launchWithExtensionKitFallback):
(WebKit::launchWithExtensionKit):
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _createVisibilityPropagationView]):
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -309,10 +316,12 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
 }
 // MARK: - WebCore::RELoaderClient
 
+#if PLATFORM(VISIONOS)
```
Should this be `PLATFORM(VISION)` (ditto for other instances in this file)?

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -351,7 +364,9 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+#if PLATFORM(VISIONOS)
     REEntitySetParent(containerEntity.get(), clientComponentEntity);
+#endif
     REEntitySetParent(m_model->rootEntity(), containerEntity.get());
```
Nit: I wonder if we should guard most of the logic to set up the layer/entity tree in this method with `PLATFORM(VISION)` for now. It's not very meaningful to parent the model entity to its container entity which is not parented to anything else anyway.

achan00 commented on Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
```diff
@@ -288,6 +292,11 @@ static String sandboxDirectory(WebCore::AuxiliaryProcessType processType, const
+    case WebCore::AuxiliaryProcessType::Model:
+        directory.append("/com.apple.WebKit.Model.Sandbox"_s);
+        break;
 #endif
```
Do we need to make extra changes to process/build model process' sandbox on the other platforms? Like `compile-sandbox.sh`, etc?
